src/t8_schemes/t8_new_feature/t8_subelements/t8_subelements_quad_cxx.cxx

```diff
@@ -0,0 +1,1325 @@
+
+  p4est_quadrant_set_morton (q, level, id);
+  T8_QUAD_SET_TDIM (q, 2);
+}
+
+t8_linearidx_t
+t8_default_scheme_sub_c::t8_element_get_linear_id (const t8_element_t *
+                                                      elem, int level)
+{
+  t8_quad_with_subelements *pquad_w_sub = (t8_quad_with_subelements *) elem;
+  p4est_quadrant_t *q = &pquad_w_sub->p4q;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+  T8_ASSERT (0 <= level && level <= P4EST_QMAXLEVEL);
+
+  return p4est_quadrant_linear_id (q, level);
+}
+
+void
+t8_default_scheme_sub_c::t8_element_first_descendant (const t8_element_t *
+                                                       elem,
+                                                       t8_element_t * desc,
+                                                       int level)
+{
+  const t8_quad_with_subelements *pquad_w_sub_elem = (const t8_quad_with_subelements *) elem;
+  t8_quad_with_subelements *pquad_w_sub_desc = (t8_quad_with_subelements *) desc;
+
+  const p4est_quadrant_t *q = &pquad_w_sub_elem->p4q;
+  p4est_quadrant_t *r = &pquad_w_sub_desc->p4q;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+  T8_ASSERT (t8_element_is_valid (desc));
+  T8_ASSERT (0 <= level && level <= P4EST_QMAXLEVEL);
+
+  p4est_quadrant_first_descendant (q, r, level);
+  T8_QUAD_SET_TDIM (r, 2);
+}
+
+void
+t8_default_scheme_sub_c::t8_element_last_descendant (const t8_element_t *
+                                                      elem,
+                                                      t8_element_t * desc,
+                                                      int level)
+{
+  const t8_quad_with_subelements *pquad_w_sub_elem = (const t8_quad_with_subelements *) elem;
+  t8_quad_with_subelements *pquad_w_sub_desc = (t8_quad_with_subelements *) desc;
+
+  const p4est_quadrant_t *q = &pquad_w_sub_elem->p4q;
+  p4est_quadrant_t *r = &pquad_w_sub_desc->p4q;
+
+  T8_ASSERT (t8_element_is_valid (elem));
+  T8_ASSERT (t8_element_is_valid (desc));
+  T8_ASSERT (0 <= level && level <= P4EST_QMAXLEVEL);
+
+  p4est_quadrant_last_descendant (q, r, level);
+  T8_QUAD_SET_TDIM (r, 2);
+}
+
+void
+t8_default_scheme_sub_c::t8_element_successor (const t8_element_t * elem1,
+                                                t8_element_t * elem2,
+                                                int level)
+{
+  const t8_quad_with_subelements *pquad_w_sub_elem1 = (const t8_quad_with_subelements *) elem1;
+  t8_quad_with_subelements *pquad_w_sub_elem2 = (t8_quad_with_subelements *) elem2;
+
+  const p4est_quadrant_t *q = &pquad_w_sub_elem1->p4q;
+  p4est_quadrant_t *r = &pquad_w_sub_elem2->p4q;
+
+  t8_linearidx_t      id;
+
+  T8_ASSERT (t8_element_is_valid (elem1));
+  T8_ASSERT (t8_element_is_valid (elem2));
+  T8_ASSERT (0 <= level && level <= P4EST_QMAXLEVEL);
+
+  id = p4est_quadrant_linear_id (q, level);
+  T8_ASSERT (id + 1 < ((t8_linearidx_t) 1) << P4EST_DIM * level);
+  p4est_quadrant_set_morton (r, level, id + 1);
+  t8_element_copy_surround (q, r);
```

Comment:
setting subelement infos shouldnt be necessary since elem2 is initialized with default values at this point and elem1 is not allowed to be a subelement (at this point). Might change in the future.